### ISSUE TITLE
Would it be possible to trigger Apple Watch complication updates through automations.

### ISSUE THREAD

**Issue by Doa132** [home-assistant/iOS#1046]
There's another app called Complications (https://mikelyons.org/complicated/) which allows you to use webhooks to send text to your apple watch complication. It's still somewhat limited to every 5-10mins, but you can choose when it happens instead of having it update randomly, potentially just missing a sensor change.
At the moment, I use it to update a watch complication only when certain sensors change by using a shell command service to send a url encoded text. Sometimes that's multiple times an hour, other times far less. Occasionally, it won't update if there's been too many sensor changes within a short time or if I haven't opened the app on my phone in ages. But it makes my complication far more reliable than is currently available through the HA iOS app complications as I'm choosing when the update occurs.
It would be great if something similar could be implemented with the HA app complications. Would be nice to be able to cut out the middle app, especially since obviously pressing on the Complicated complication can't open the HA actions.

**Comment by zacwest** [comment]
It is possible to update complications via a push notification, but one potential downside is going to be how often you can do it. The complications are limited to 50 total refreshes per day. That can be once per second for 50 seconds, total, or once every 30 minutes for 24 hours, or so. Right now the app prefers the latter.

**Comment by Doa132** [comment]
Even with the limit of 50 per day it could be more useful to have it change based on a sensor change. For example triggering a change only if your alarm system changes state. Mostly like wouldn’t happen often, but better to be able to see that it’s disarmed before you try open your door than to have it randomly update 20mins later when the state hasn’t changed. 
It would just have to be made clear in the docs that there is a limit to how often it will update.

**Comment by zacwest** [comment]
Looks like this will require setting up a notify service specifically for the watch for watchOS 6+. In my reading of [the docs](https://developer.apple.com/documentation/clockkit/updating_your_complication#) the 50/day limit for push notifications may be separate from the periodic background uploading we do.

**Comment by zacwest** [comment]
OK, so I've got this working locally, but it's going to be relatively complicated:

1. This requires using PushKit, which is a separate (but compatible) push system to the normal APNs we use to deliver alerts.
2. PushKit actually generates new device identifiers for just complication notifications. 
3. Firebase doesn't support PushKit notifications. That is, they don't allow you to just push to arbitrary Apple tokens and they have not done the effort to wrap PushKit.

A good amount of infrastructure work is going to be necessary to get this working, unfortunately.

**Comment by zacwest** [comment]
for later, this registers with PushKit. the registration request also has credentials with the token in it as Data which is the device token to push to

```
diff --git a/Sources/Extensions/Watch/ExtensionDelegate.swift b/Sources/Extensions/Watch/ExtensionDelegate.swift
index 3c909ade..d69dccb7 100644
--- a/Sources/Extensions/Watch/ExtensionDelegate.swift
+++ b/Sources/Extensions/Watch/ExtensionDelegate.swift
@@ -14,6 +14,7 @@ import UserNotifications
 import XCGLogger
 import Shared
 import PromiseKit
+import PushKit
 
 class ExtensionDelegate: NSObject, WKExtensionDelegate {
     // MARK: Fileprivate
@@ -24,6 +25,8 @@ class ExtensionDelegate: NSObject, WKExtensionDelegate {
         }
     }
 
+    var registry: AnyObject?
+
     // MARK: - WKExtensionDelegate -
 
     func applicationDidFinishLaunching() {
@@ -48,6 +51,13 @@ class ExtensionDelegate: NSObject, WKExtensionDelegate {
 
         setupWatchCommunicator()
 
+        if #available(watchOSApplicationExtension 6.0, *) {
+            registry = with(PKPushRegistry(queue: nil)) {
+                $0.desiredPushTypes = [.complication]
+                $0.delegate = self
+            }
+        }
+
         // schedule the next background refresh
         Current.backgroundRefreshScheduler.schedule().cauterize()
     }
@@ -232,6 +242,16 @@ class ExtensionDelegate: NSObject, WKExtensionDelegate {
         }
     }
 
+    @available(watchOSApplicationExtension 6.0, *)
+    func didReceiveRemoteNotification(
+        _ userInfo: [AnyHashable: Any],
+        fetchCompletionHandler completionHandler: @escaping (WKBackgroundFetchResult) -> Void
+    ) {
+        updateComplications().done {
+            completionHandler(.newData)
+        }
+    }
+
     func didRegisterForRemoteNotifications(withDeviceToken deviceToken: Data) {
         let apnsToken = deviceToken.map { String(format: "%02.2hhx", $0) }.joined()
         Current.Log.verbose("Successfully registered for push notifications! APNS token: \(apnsToken)")
@@ -249,3 +269,17 @@ extension ExtensionDelegate: UNUserNotificationCenterDelegate {
         completionHandler([.alert, .badge, .sound])
     }
 }
+
+@available(watchOSApplicationExtension 6.0, *)
+extension ExtensionDelegate: PKPushRegistryDelegate {
+    func pushRegistry(_ registry: PKPushRegistry, didUpdate pushCredentials: PKPushCredentials, for type: PKPushType) {
+        Current.Log.info("push credentials \(pushCredentials) updated for type \(type)")
+    }
+
+    func pushRegistry(_ registry: PKPushRegistry, didReceiveIncomingPushWith payload: PKPushPayload, for type: PKPushType, completion: @escaping () -> Void) {
+        Current.Log.info("received incoming push: \(payload) for \(type)")
+        updateComplications().done {
+            completion()
+        }
+    }
+}
```
